src/SAWScript/Builtins.hs

```diff
@@ -892,11 +897,90 @@ offline_extcore path = proveWithExporter (const Prover.writeCoreProp) path ".ext
 offline_smtlib2 :: FilePath -> ProofScript SV.SatResult
 offline_smtlib2 path = proveWithExporter Prover.writeSMTLib2 path ".smt2"
 
+w4_offline_smtlib2 :: FilePath -> ProofScript SV.SatResult
+w4_offline_smtlib2 path = proveWithExporter Prover.writeSMTLib2What4 path ".smt2"
+
 offline_unint_smtlib2 :: [String] -> FilePath -> ProofScript SV.SatResult
 offline_unint_smtlib2 unints path =
   do unintSet <- lift $ resolveNames unints
      proveWithExporter (Prover.writeUnintSMTLib2 unintSet) path ".smt2"
 
+offline_verilog :: FilePath -> ProofScript SV.SatResult
+offline_verilog path =
+  proveWithExporter (Prover.adaptExporter Prover.write_verilog) path ".v"
+
+parseAigerCex :: String -> IO [Bool]
+parseAigerCex text =
+  case lines text of
+    [_, cex] ->
+      case words cex of
+        [bits, _] -> mapM bitToBool bits
+        _ -> fail $ "invalid counterexample line: " ++ cex
+    _ -> fail $ "invalid counterexample text: " ++ text
+  where
+    bitToBool '0' = return False
+    bitToBool '1' = return True
+    bitToBool c   = fail ("invalid bit: " ++ [c])
+
+w4_abc_verilog :: ProofScript SV.SatResult
+w4_abc_verilog = wrapW4Prover w4AbcVerilog []
+
+w4AbcVerilog ::
+  Set VarIndex ->
+  SharedContext ->
+  Bool ->
+  Prop ->
+  IO (Maybe [(String, FirstOrderValue)], SolverStats)
+w4AbcVerilog _unints sc _hashcons g =
+       -- Create temporary files
+    do let tpl = "abc_verilog.v"
+           tplCex = "abc_verilog.cex"
+       tmp <- emptySystemTempFile tpl
+       tmpCex <- emptySystemTempFile tplCex
+
+       -- Get goal into the right form
+       let (goalArgs, concl) = asPiList (unProp g)
+       p <- case asEqTrue concl of
+              Just p -> return p
+              Nothing -> fail "adaptExporter: expected EqTrue"
+       t <- scLambdaList sc goalArgs =<< scNot sc p
+       Prover.writeVerilog sc tmp t
+
+       -- Run ABC and remove temporaries
+       let execName = "abc"
+           args = ["-q", "%read " ++ tmp ++"; %blast; &sweep -C 5000; &syn4; &cec -m; write_aiger_cex " ++ tmpCex]
+       (ec, out, err) <- readProcessWithExitCode execName args ""
+       cexText <- readFile tmpCex
+       removeFile tmp
+       removeFile tmpCex
+
+       -- Parse and report results
+       let isEquivalent = "equivalent" `isInfixOf` out
+           isNotEquivalent = "NOT EQUIVALENT" `isInfixOf` out
+           stats = solverStats "abc_verilog" (scSharedSize (unProp g))
+       res <- case (isEquivalent, isNotEquivalent) of
+                (True, False) -> return $ Nothing
+                (False, True) ->
+                  do bits <- reverse <$> parseAigerCex cexText
+                     let goalArgs' = reverse goalArgs
+                         argTys = map snd goalArgs'
+                         argNms = map fst goalArgs'
+                         r = liftCexBB (mapMaybe asFiniteTypePure argTys) bits
```

Comment:
It's probably best to avoid `asFiniteTypePure` in favor of `asFiniteType`. It looks like the surrounding code is in `IO`, so you don't really need to use the pure variant. Compared to `asFiniteType`, `asFiniteTypePure` skips the term normalization step, so it can fail in situations where `asFiniteType` wouldn't.

Also, it seems like this use of `mapMaybe` is just wrong; is the desired behavior really to just filter out those arguments for which `asFiniteTypePure` fails? Perhaps the whole thing should actually fail if any of the argument types are invalid?

Reply:
Yeah, having the whole thing fail for any invalid argument type seems like the right behavior. I'll do that.